### PR TITLE
XComposite: Fix capturing on software rasterization setups

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
### Description
The GLX_EXT_texture_from_pixmap spec says:

> The contents of the texture after the drawable has been bound are defined
> as the result of all rendering that has completed before the call to
> glXBindTexImageEXT.  In other words, the results of any operation which
> has caused damage on the drawable prior to the glXBindTexImageEXT call
> will be represented in the texture.
> 
> Rendering to the drawable while it is bound to a texture will leave the
> contents of the texture in an undefined state.  However, no
> synchronization between rendering and texturing is done by GLX.  It is
> the application's responsibility to implement any synchronization
> required.

In practice, on most systems with a GPU this kept a directly binding to
the framebuffer, which made it work with the previous code. However,
on software rasterization setups using llvmpipe, a blit was done inside
the call to glXBindTexImageEXT. This was the cause of a notoriously
ignored bug where the captured image would "freeze" until the source
was reconfigured.

### Motivation and Context

Example complaint:

https://obsproject.com/forum/threads/xcomposite-window-capture-is-frozen-does-not-update.136729/

### How Has This Been Tested?

This has been tested on a xpra virtual desktop, which uses llvmpipe unless configured otherwise.

### Types of changes

- Bug fix (non-breaking change which fixes an issue)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- N/A I have included updates to all appropriate documentation.

### CODE REVIEW

File: plugins/linux-capture/xcompcap-main.cpp [hunk 1]
```diff
@@ -135,7 +135,7 @@ void XCompcapMain::defaults(obs_data_t *settings)
 	obs_data_set_default_bool(settings, "exclude_alpha", false);
 }
 
-#define FIND_WINDOW_INTERVAL 0.5
```
Comment (kkartaltepe) [inline plugins/linux-capture/xcompcap-main.cpp:138]: Is this change related to the fix?

File: plugins/linux-capture/xcompcap-main.cpp [hunk 2]
```diff
@@ -621,6 +621,7 @@ void XCompcapMain::tick(float seconds)
 	}
 
 	XDisplayLock xlock;
+	XErrorLock xerrlock;
```
Comment (kkartaltepe) [inline plugins/linux-capture/xcompcap-main.cpp:624]: we are already holding an xserver lock in XDisplayLock, did you mean to replace this with the XErrorLock?

**Comment by ezette** [comment]
Do not know

File: plugins/linux-capture/xcompcap-main.cpp (hunk 1 above)
Comment (ishitatsuyuki) [reply plugins/linux-capture/xcompcap-main.cpp]: Oops, I accidentally pushed some debug code. It is now removed.

File: plugins/linux-capture/xcompcap-main.cpp (hunk 2 above)
Comment (ishitatsuyuki) [reply plugins/linux-capture/xcompcap-main.cpp]: Yes, I didn't realize a XErrorLock supersedes a XDisplayLock. Thanks!

**Comment by norihiro** [comment]
The commit title should start from `linux-capture:` instead of `XComposite:`.

**Comment by ishitatsuyuki** [comment]
Fixed.

**Comment by kkartaltepe** [comment]
I have two questions,

1) now that we bind the texture do we we need to move the release into the rendering loop. The recommended calls are 
```
XGrabServer()
glXWaitX() or XSync()
glXBindTexImageEXT()

<Do rendering/compositing>

glXReleaseTexImageEXT()
XUngrabServer()
```
In the extension. Will we be leaking something by repeatedly binding without releasing?

2) Can we avoid the log spam when binding fails, such as for offscreen windows? This change currently introduces an error in the log per frame in this case.

**Comment by jp9000** [comment]
Hey there, just wanted to check back with you, we just had those few remaining questions there from @kkartaltepe.

**Comment by ishitatsuyuki** [comment]
Sorry, couldn't prioritize this due to other work but I'll try to get back on this soon.

**Comment by ishitatsuyuki** [comment]
I have updated the patch.

The bind and release calls are now matched, in order to avoid causing potentially implementation dependent leaks. It also binds the correct texture before `glXBindTexImageEXT` now.

Only the first error from `tick` will be logged in order to avoid spam. Logging is re-enabled on each reconfiguration.

**Comment by kkartaltepe** [comment]
Thanks this looks good on my machine, if you get a chance and can apply this patch that would be nice otherwise ill probably make a PR for it later. (Just avoiding stereoscopic framebuffer name and aesthetic changes to the tick error check).

```diff
--- a/plugins/linux-capture/xcompcap-main.cpp
+++ b/plugins/linux-capture/xcompcap-main.cpp
@@ -549,7 +549,7 @@ void XCompcapMain::updateSettings(obs_data_t *settings)
                                     GS_GL_DUMMYTEX);
        GLuint gltex = *(GLuint *)gs_texture_get_obj(p->gltex);
        glBindTexture(GL_TEXTURE_2D, gltex);
-       glXBindTexImageEXT(xdisp, p->glxpixmap, GLX_FRONT_LEFT_EXT, nullptr);
+       glXBindTexImageEXT(xdisp, p->glxpixmap, GLX_FRONT_EXT, nullptr);
        if (xlock.gotError()) {
                blog(LOG_ERROR, "glXBindTexImageEXT failed: %s",
                     xlock.getErrorText().c_str());
@@ -563,7 +563,7 @@ void XCompcapMain::updateSettings(obs_data_t *settings)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
        // glxBindTexImageEXT might modify the textures format.
        gs_color_format format = gs_format_from_tex();
-       glXReleaseTexImageEXT(xdisp, p->glxpixmap, GLX_FRONT_LEFT_EXT);
+       glXReleaseTexImageEXT(xdisp, p->glxpixmap, GLX_FRONT_EXT);
        if (xlock.gotError()) {
                blog(LOG_ERROR, "glXReleaseTexImageEXT failed: %s",
                     xlock.getErrorText().c_str());
@@ -653,13 +653,11 @@ void XCompcapMain::tick(float seconds)
        }
 
        glBindTexture(GL_TEXTURE_2D, *(GLuint *)gs_texture_get_obj(p->gltex));
-       glXBindTexImageEXT(xdisp, p->glxpixmap, GLX_FRONT_LEFT_EXT, nullptr);
-       if (xlock.gotError()) {
-               if (!p->tick_error_suppressed)
-                       blog(LOG_ERROR, "glXBindTexImageEXT failed: %s",
-                            xlock.getErrorText().c_str());
+       glXBindTexImageEXT(xdisp, p->glxpixmap, GLX_FRONT_EXT, nullptr);
+       if (xlock.gotError() && !p->tick_error_suppressed) {
+               blog(LOG_ERROR, "glXBindTexImageEXT failed: %s",
+                    xlock.getErrorText().c_str());
                p->tick_error_suppressed = true;
-               xlock.resetError();
        }
 
        if (p->include_border) {
@@ -672,13 +670,11 @@ void XCompcapMain::tick(float seconds)
                                       height());
        }
 
-       glXReleaseTexImageEXT(xdisp, p->glxpixmap, GLX_FRONT_LEFT_EXT);
-       if (xlock.gotError()) {
-               if (!p->tick_error_suppressed)
-                       blog(LOG_ERROR, "glXReleaseTexImageEXT failed: %s",
-                            xlock.getErrorText().c_str());
+       glXReleaseTexImageEXT(xdisp, p->glxpixmap, GLX_FRONT_EXT);
+       if (xlock.gotError() && !p->tick_error_suppressed) {
+               blog(LOG_ERROR, "glXReleaseTexImageEXT failed: %s",
+                    xlock.getErrorText().c_str());
                p->tick_error_suppressed = true;
-               xlock.resetError();
        }
        glBindTexture(GL_TEXTURE_2D, 0);
 
```

**Comment by ishitatsuyuki** [comment]
Applied. Just FYI, I had to ignore whitespaces when applying the patch and convert the indents to tab later.

**Comment by kkartaltepe** [comment]
Ah right whitespace gets ruined in github, my bad.

**Comment by univrsal** [comment]
Just a heads up, it seems like this is causing performance issues on proprietary nvidia drivers.

Log for 4623a6b4bc4c89ceb5db684e2e7fbd57d01129aa (no rendering lag):
https://obsproject.com/logs/GMoA70ZoAVngkg9S
Log for 316f858c68d3ef3a73166191b6f89b7bc3f3387c:
https://obsproject.com/logs/AqYbf16E3VT5g6cw

For me it just reduced the fps from 60 to 30 when showing a window capture. When moving a window fps would go down even further to around 15. When I checked out the previous commit things went back to normal.

**Comment by ishitatsuyuki** [comment]
@univrsal Can you create an issue? I think this needs more investigation to resolve, and even should the PR be reverted, an issue is needed to track the investigations and progress before this can reland.
